Fix SI-8202 and improve support for splicing patterns into vals

densh opened the pull request:
Fixes:
- SI-8202 makes quasiquote parse q"val (x: Int) = 1" the same as q"val x: Int = 1" 
- Splicing of patterns into vals which was quite broken previously

review @xeno-by @retronym

densh committed:
SI-8202 bug compatibility with SI-8211 for quasiquotes

densh commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -33,10 +33,10 @@ trait Printers extends api.Printers { self: SymbolTable =>
+    if (sym == null || sym == NoSymbol)
       qname
+    else if (sym.name.toTermName == nme.ERROR)
+      s"<$qname: error>"
```
Used to NPE on me before the change.

xeno-by replied:
Use `sym.isErroneous` instead

densh replied:
That method is not available there.

xeno-by replied:
What do you mean? It's reflect/internal/Printers.

densh replied:
Must have been a glitch, I've used it there and had compilation error. Not anymore.

densh commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -1093,7 +1093,7 @@ trait Printers extends api.Printers { self: SymbolTable =>
           print("pendingSuperCall")
         case tree: Tree =>
-          val hasSymbolField = tree.hasSymbolField && tree.symbol != NoSymbol
+          def hasSymbolField = tree.hasSymbolField && tree.symbol != NoSymbol && tree.symbol != null
```
This one [too](https://github.com/scala/scala/pull/3455/files#r9370143).

xeno-by replied:
If `hasSymbolField` is true then `tree.symbol` should not be null. Could you elaborate on when it used to crash?

densh replied:
I've been debugging some quasiquote code with -Yquasiquote-debug and it crashed there with NPE.

xeno-by replied:
Learning the details would help greatly, as there is compiler code that relies on the fact that `hasSymbolField == true` means non-null return value from `symbol`.

densh replied:
1. Remove null check on top
2. Start repl with -Yquasiquote-debug
3. Type:

```
scala> val q"{ ..$stats1 }" = q" "
...
<console>:29: error: exception during macro expansion:
java.lang.NullPointerException
    at scala.reflect.internal.Printers$RawTreePrinter$$anonfun$print$3.apply(Printers.scala:1097)
    at scala.reflect.internal.Printers$RawTreePrinter$$anonfun$print$3.apply(Printers.scala:1082)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.reflect.internal.Printers$RawTreePrinter.print(Printers.scala:1082)
    at scala.reflect.internal.Printers$RawTreePrinter$$anonfun$print$3$$anonfun$apply$9.apply(Printers.scala:1131)
    at scala.reflect.internal.Printers$RawTreePrinter$$anonfun$print$3$$anonfun$apply$9.apply(Printers.scala:1105)
    at scala.reflect.internal.Printers$RawTreePrinter.printIterable(Printers.scala:1205)
    at scala.reflect.internal.Printers$RawTreePrinter.printProduct(Printers.scala:1191)
    at scala.reflect.internal.Printers$RawTreePrinter$$anonfun$print$3.apply(Printers.scala:1098)
    at scala.reflect.internal.Printers$RawTreePrinter$$anonfun$print$3.apply(Printers.scala:1082)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
    at scala.reflect.internal.Printers$RawTreePrinter.print(Printers.scala:1082)
    at scala.reflect.api.Printers$class.render(Printers.scala:181)
    at scala.reflect.api.Universe.render(Universe.scala:61)
    at scala.reflect.api.Printers$class.showRaw(Printers.scala:229)
    at scala.reflect.api.Universe.showRaw(Universe.scala:61)
    at scala.tools.reflect.quasiquotes.Quasiquotes$$anonfun$expandQuasiquote$3.apply(Quasiquotes.scala:49)
    at scala.tools.reflect.quasiquotes.Quasiquotes$$anonfun$expandQuasiquote$3.apply(Quasiquotes.scala:49)
    at scala.tools.reflect.quasiquotes.Quasiquotes.debug(Quasiquotes.scala:15)
    at scala.tools.reflect.quasiquotes.Quasiquotes.expandQuasiquote(Quasiquotes.scala:49)
    at scala.tools.reflect.FastTrack$$anonfun$1$$anonfun$apply$8$$anonfun$applyOrElse$8.apply(FastTrack.scala:54)
    at scala.tools.reflect.FastTrack$$anonfun$1$$anonfun$apply$8$$anonfun$applyOrElse$8.apply(FastTrack.scala:54)
    at scala.tools.reflect.FastTrack$FastTrackEntry.apply(FastTrack.scala:37)
    at scala.tools.reflect.FastTrack$FastTrackEntry.apply(FastTrack.scala:32)
    at scala.tools.nsc.typechecker.Macros$class.macroExpandWithRuntime(Macros.scala:739)

       val q"{ ..$stats1 }" = q" "
```

densh replied:
I've poked around and it looks like current master isn't crashing like that. The only change I can see being somehow related is introduction of `implodePatDefs` transformer that preprocesses trees from the parser. Is there anything there that can cause `symbol` being set to `null`?

retronym replied:
If you can't imagine your way to the cuplable code, try this:

Temporarily change `SymTree#symbol_=` to keep track of all modifications, rather than overrwrite the field. When you hit spot the tree with the null symbol later on, print out the list of places that called `symbol_=`. (You can add `Thread.currentThread.getStackTrace` to a buffer. Just make sure you have a big heap! You might also need to change `TreeCopier` to propagate your debug info.

densh replied:
It turned that transformer was indeed the cause of all trouble. I've used `SyntacticBlock` in combination with `copyAttrs` not taking into consideration the fact that `SyntacticBlocks` can collapse on single-element input causing `copyAttrs` copy stuff into a wrong tree.

xeno-by commented:
Why not parse `val (x: Int) = 1` as a type match pattern? There was also one JIRA issue updated recently that could be fixed by that.

densh commented:
@xeno-by I'd love to see that happen. But it looks like there are some concerns about change of behavior in existing code. We need to discuss it one of next week's meetings.

xeno-by commented:
@densh What are the concerns? Could you comment? /cc @retronym @adriaanm

densh commented:
@retronym doesn't like the fact that code like:

``` scala
val (x: Int) = 1
```

Will change meaning from 2.10 to 2.11 (compile-time check will become run-time check.) It will be 100% source compatible though and I doubt that anyone uses such a form anyway considering the fact that it's exactly the same as the one without parens atm.

densh commented:
ping @xeno-by

xeno-by commented:
What are the performance implications of this pull request?

densh commented:
After discussion today the conclusion was that the behavior of `val (x: T) = ...` will not change in 2.11 but should be discussed again in 2.12 cycle. Therefore fix that breaks quasiquotes to be broken in the same way as the parser is appropriate here.

adriaanm commented:
Thinking about it some more, it seems we can likely just fix this in 2.12 outright, ideally with an opt-in warning that lets users detect where their program is affected. (As part of the style/lint  checker)

adriaanm commented:
(Where "this" is consistent interpretation of tuple patterns in val and for)

densh commented:
@xeno-by what kind of performance are you talking about? Runtime isn't affected in any way. Compile-time of the quasiquote will perform implodePatDefs transform on every parsed tree.

xeno-by commented:
@densh I see. Thanks

densh commented:
@adriaanm Sounds good. Looking forward too it.

densh committed:
Move placeholder construction logic into Placeholders

Previously construction logic used to be in Parsers and deconstruction in
Placeholders making it easy to forget one if you change the other.

densh committed:
Move null check case higher to avoid NPE

Also use sym.isErrorneous instead of manual name check.

densh committed:
Improve support for patterns in vals

This commits adds construction-only support for splicing patterns into
vals (a.k.a. PatDef). Due to non-locality of the desugaring it would
have been quite expensive to support deconstruction as the only way to
do it with current trees is to perform implodePatDefs transformation on
every single tree.

densh commented:
@xeno-by Just rebased and added some extra tests to support `q"val $pat: $tpt = $rhs"` case and minor refactoring of placeholders.

densh commented:
ping @xeno-by,  please merge early to prevent rebasing issues with reflection pr

retronym commented:
LGTM
